File: software/examples/go/example_callback.go
package main

import (
	"fmt"
	"github.com/Tinkerforge/go-api-bindings/energy_monitor_bricklet"
	"github.com/Tinkerforge/go-api-bindings/ipconnection"
)

const ADDR string = "localhost:4223"
const UID string = "XYZ" // Change XYZ to the UID of your Energy Monitor Bricklet.

func main() {
	ipcon := ipconnection.New()
	defer ipcon.Close()
	em, _ := energy_monitor_bricklet.New(UID, &ipcon) // Create device object.

	ipcon.Connect(ADDR) // Connect to brickd.
	defer ipcon.Disconnect()
	// Don't use device before ipcon is connected.

	em.RegisterEnergyDataCallback(func(voltage int32, current int32, energy int32, realPower int32, apparentPower int32, reactivePower int32, powerFactor uint16, frequecy uint16) {
		fmt.Printf("Voltage: %f V\n", float64(voltage)/100.0)
		fmt.Printf("Current: %f A\n", float64(current)/100.0)
		fmt.Printf("Energy: %f Wh\n", float64(energy)/100.0)
		fmt.Printf("Real Power: %f h\n", float64(realPower)/100.0)
		fmt.Printf("Apparent Power: %f VA\n", float64(apparentPower)/100.0)
		fmt.Printf("Reactive Power: %f VAR\n", float64(reactivePower)/100.0)
		fmt.Printf("Power Factor: %f\n", float64(powerFactor)/1000.0)
		fmt.Printf("Frequency: %f Hz\n", float64(frequecy)/100.0)
		fmt.Println()
	})

	// Set period for Energy Data callback to 1s (1000ms).
	em.SetEnergyDataCallbackConfiguration(1000, false)

	fmt.Print("Press enter to exit.")
	fmt.Scanln()
}

File: software/examples/rust/example_callback.rs
use std::{error::Error, io, thread};
use tinkerforge::{energy_monitor_bricklet::*, ip_connection::IpConnection};

const HOST: &str = "localhost";
const PORT: u16 = 4223;
const UID: &str = "XYZ"; // Change XYZ to the UID of your Energy Monitor Bricklet.

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection.
    let em = EnergyMonitorBricklet::new(UID, &ipcon); // Create device object.

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd.
                                          // Don't use device before ipcon is connected.

    let energy_data_receiver = em.get_energy_data_callback_receiver();

    // Spawn thread to handle received callback messages.
    // This thread ends when the `em` object
    // is dropped, so there is no need for manual cleanup.
    thread::spawn(move || {
        for energy_data in energy_data_receiver {
            println!("Voltage: {} V", energy_data.voltage as f32 / 100.0);
            println!("Current: {} A", energy_data.current as f32 / 100.0);
            println!("Energy: {} Wh", energy_data.energy as f32 / 100.0);
            println!("Real Power: {} h", energy_data.real_power as f32 / 100.0);
            println!("Apparent Power: {} VA", energy_data.apparent_power as f32 / 100.0);
            println!("Reactive Power: {} VAR", energy_data.reactive_power as f32 / 100.0);
            println!("Power Factor: {}", energy_data.power_factor as f32 / 1000.0);
            println!("Frequency: {} Hz", energy_data.frequency as f32 / 100.0);
            println!();
        }
    });

    // Set period for Energy Data callback to 1s (1000ms).
    em.set_energy_data_callback_configuration(1000, false);

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    ipcon.disconnect();
    Ok(())
}
